File: design-pattern/adapter-pattern/adapterPatter.py
class Target:
    """
    ターゲットインターフェース(抽象)
    """

    def request(self):
        pass


class Adaptee:
    """
    既存クラス(具象)
    """

    def specific_request(self):
        print("既存クラスのメソッドが呼ばれました。")


class Adapter(Target):
    """
    アダプタークラス(具象)
    Targetインタフェースを実装する
    """

    def __init__(self, adaptee: Adaptee):
        # コンストラクタで、既存クラスのインスタンスを受け取る
        self.adaptee = adaptee

    def request(self):
        self.adaptee.specific_request()


if __name__ == '__main__':
    adaptee = Adaptee()
    adapter = Adapter(adaptee)
    adapter.request()

File: design-pattern/adapter-pattern/adapterPatter.cpp
/*
このコードでは、Target クラスがターゲットインターフェースを定義し、Adapteeクラスがアダプティークラス（既存のクラス）を定義しています。
Adapter クラスがアダプタークラスを定義し、Target インターフェースを実装します。
アダプタークラスのコンストラクタで、アダプティークラスのポインタを受け取ります。
Adapter クラスの request() メソッドが呼ばれた場合、アダプティークラスの specificRequest() メソッドを呼び出します。
最後に、Adaptee クラスの specificRequest() メソッドが呼ばれることを確認するために、Adapter クラスのインスタンスを作成し、request() メソッドを呼び出します。
このコードを実行すると、以下のように出力されます。
*/

#include <iostream>

class Target
{
public:
    virtual void request() = 0;
};

class Adaptee
{
public:
    void specificRequest()
    {
        std::cout << "アダプティーのメソッドが呼ばれました。" << std::endl;
    }
};

class Adapter : public Target
{
private:
    Adaptee *adaptee;

public:
    Adapter(Adaptee *adaptee)
    {
        this->adaptee = adaptee;
    }

    void request()
    {
        adaptee->specificRequest();
    }
};

int main()
{
    Adaptee *adaptee = new Adaptee();
    Target *adapter = new Adapter(adaptee);
    adapter->request();
    return 0;
}
